src/shims/fs.rs

```diff
@@ -154,27 +166,51 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         this.check_no_isolation("read")?;
 
-        let count = this.read_scalar(count_op)?.to_machine_usize(&*this.tcx)?;
+        let ptr_size = this.pointer_size().bits();
+
+        // We cap the number of read bytes to the largest value that we are able to fit in both the
+        // host's and target's `isize`.
+        let count = this
+            .read_scalar(count_op)?
+            .to_machine_usize(&*this.tcx)?
+            .min(1 << (ptr_size - 1))
```

Comment:
```suggestion
            .min((1 << (ptr_size - 1)) - 1) // max value of target isize
```
We should probably have helper methods for this somewhere... but that's for later.

Reply:
Yes I thought the same.